lib/plugins/backup.rb

```diff
@@ -0,0 +1,168 @@
+        end
+      end
+
+      # Main worker method, creates a tarball of the given
+      # files, and uploads to an S3 bucket.
+      # @param resources [Resources] injected Moonshot resources
+      def backup(resources) # rubocop:disable Metrics/AbcSize
+        raise ArgumentError if resources.nil?
+
+        @app_name = resources.stack.app_name
+        @stack_name = resources.stack.name
+        @target_name = render(@target_name)
+
+        resources.ilog.start log_message.clone << ' in progress.' do |s|
+          begin
+            tar_out = tar @files
+            zip_out = zip tar_out
+            upload zip_out
+
+            s.success log_message.clone << ' succeeded.'
+          rescue StandardError => e
+            s.failure log_message.clone << " failed: #{e}"
+          ensure
+            tar_out.close unless tar_out.nil?
+            zip_out.close unless zip_out.nil?
+          end
+        end
+      end
+
+      # Dynamically responding to hooks
+      # supplied in the constructor
+      def method_missing(method_name, *args, &block)
+        @hooks.include?(method_name) ? backup(*args) : super
+      end
+
+      def respond_to?(method_name, include_private = false)
+        @hooks.include?(method_name) || super
+      end
+
+      private
+
+      attr_accessor :app_name,
+                    :stack_name
+
+      # Creates an IO stream from the passed files
+      # @param target_files [String] array of filenames
+      # @return tar_stream [IO] IO stream of archived files
+      def tar(target_files)
+        tar_stream = StringIO.new
+        Gem::Package::TarWriter.new(tar_stream) do |writer|
+          target_files.each do |file|
+            file = render_path file
+            writer.add(file)
+          end
+        end
+        tar_stream.seek(0)
+        tar_stream
+      end
+
+      # Zips up the passed tar stream
+      # @param io_tar [IO] tar stream
+      # @return zip_stream [IO] IO stream of zipped file
+      def zip(io_tar)
+        zip_stream = StringIO.new
+        Zlib::GzipWriter.wrap(zip_stream) do |gz|
+          gz.write(io_tar.read)
+          gz.finish
+        end
+        zip_stream.seek(0)
+        zip_stream
+      end
+
+      # Uploads an object from the pass
+      # IO stream to the specified bucket
+      # @param io_zip [IO] tar stream
+      def upload(io_zip)
+        s3_client.put_object(
+          acl: 'private',
+          bucket: @bucket,
+          key: @target_name,
+          body: io_zip
+        )
+      end
+
+      # Renders string with the specified placeholders
+      # @param io_zip [String] raw string with placeholders
+      # @return [String] rendered string
+      def render(placeholder)
+        format(
+          placeholder,
+          app_name: @app_name,
+          stack_name: @stack_name,
+          timestamp: Time.now.to_i.to_s,
+          user: ENV['USER']
+        )
+      end
+
+      # Renders string with the specified placeholders
```

Comment:
I think if you switch to an array of paths, this method just goes away and becomes something like:
```ruby
paths = paths.map { |p| render(p) }
```